contracts/RemmeBridge.sol

```diff
@@ -1,10 +1,240 @@
+    function getSwapInfo(bytes32 _swapId) view returns (
+        address sender,
+        address receiver,
+        address keyHolder,
+        bytes remChainAddress,
+        State state)
+    {
+        AtomicSwap storage swap = swaps[_swapId];
+        sender = swap.senderAddress;
+        receiver = swap.receiverAddress;
+        keyHolder = swap.keyHolderAddress;
+        remChainAddress = swap.remchainAddress;
+        state = swapStates[_swapId];
+    }
+
+    /// @notice Get swap details info
+    function getSwapDetails(bytes32 _swapId) view returns (
+        uint amount,
+        bytes emailEncrypted,
+        bytes32 secretLock,
+        bytes32 secretKey,
+        uint timelock)
+    {
+	    AtomicSwap storage swap = swaps[_swapId];
+        amount = swap.amount;
+        emailEncrypted = swap.emailAddressEncryptedOptional;
+        secretLock = swap.secretLock;
+        secretKey = swap.secretKey;
+        timelock = swap.timelock;
+    }
+
+    /*
+	* @notice User should use this function for request swap locking his tokens
+	* @dev There are two cases of using this function:
+	* 1) Alice requests swap and don't put _secretLock argument, and sends ether
+	*    required by atomicSwapProvider for gas coverage
+	* 2) Bob opens swap after Alice's request in Remmechain and set _secretLock
+	* @param _secretLock (!) will be set only by Bob. Should be checked in client side during validation
+    */
+    function openSwap (
+        bytes32 _swapId,
+        address _receiverAddress,
+        bytes32 _secretLock,
+        uint256 _amount,
+        bytes _remchainAddress,
+        bytes _emailAddressEncryptedOptional)
+    external
+    {
+	    require(swaps[_swapId].swapId == 0);
+
+        //Bob always set as keyHolder
+	    uint256 lock;
+	    address keyHolder;
+	    uint256 amountToSwap = _amount;
+        if (_secretLock == bytes32(0)) {
+            lock = now + LOCK24;
+            keyHolder = _receiverAddress;
+            if (_receiverAddress == atomicSwapProvider) {
+                require(REMToken.transferFrom(msg.sender, coldStorage, providerFee));
+	            amountToSwap = _amount.sub(providerFee);
+            }
+        } else {
+            lock = now + LOCK48;
+            keyHolder = msg.sender;
+        }
+
+	    AtomicSwap memory swap = AtomicSwap({
+            swapId: _swapId,
+            senderAddress: msg.sender,
+            receiverAddress: _receiverAddress,
+            keyHolderAddress: keyHolder,
+            remchainAddress: _remchainAddress,
+            amount: amountToSwap,
+            emailAddressEncryptedOptional: _emailAddressEncryptedOptional,
+            secretLock: _secretLock,
+            secretKey: "", //will be set only on closing swap
+            timelock: lock
+            });
+        swaps[swap.swapId] = swap;
+
+	    require(REMToken.transferFrom(msg.sender, address(this), amountToSwap));
+
+        swapStates[swap.swapId] = State.OPENED;
```

Comment:
Is a separate variable really required?

Reply:
Yes, contract logic is based on states for incorrect user actions prevention

Reply:
In that case there is 3 related extra lines in open swap needed

Reply:
yeah, at least from the beginning one already will know it is APPROVED

Reply:
ok, we can change it